src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java

```diff
@@ -1,85 +1,149 @@
     }
 
-    public void enableHopper() {
-        this.hopperController.enableCurrentLimit(true);
-        this.hopperController.set(0.8d);
+    public void task() {
+        switch(this.state) {
+            case PRIMING:
+                if(this.topUltrasonic.getRangeMM() > this.topUltrasonicDistanceMM) {
+                    this.enable(FeederComponent.HOPPER);
+                    this.enable(FeederComponent.CONVEYOR);
+                } else {
+                    this.state = FeederState.OFF;
+                }
+                break; // It will behave as expected because this runs in a loop
+            case FEEDING:
+                this.enable(FeederComponent.HOPPER);
+                if(this.shooting){
+                    this.enable(FeederComponent.CONVEYOR);
+                } else {
+                    this.stop(FeederComponent.CONVEYOR);
+                }
+                break;
+            case UNLOADING:
+                if(!this.topPoller.checkUnloaded()) {
+                    this.enable(FeederComponent.HOPPER, FeederDirection.REVERSE);
+                    this.enable(FeederComponent.CONVEYOR, FeederDirection.REVERSE);
+                    break;
+                } else {
+                    this.state = FeederState.OFF;
+                }
+            default:
+                
+        }
     }
 
-    public void stopConveyor() {
-        this.beltController.set(0.0d);
+    @Override
+    public void feed() {
+        this.state = FeederState.FEEDING;
+    }
 
+    @Override
+    public void defineResources() {
+        require(this.topUltrasonic);
     }
 
-    public void stopHopper() {
-        this.hopperController.enableCurrentLimit(false);
-        this.hopperController.set(0.0d);
+    @Override
+    public void prime() {
+        this.state = FeederState.PRIMING;
     }
 
-    public void stopAll() {
-        //this.beltController.set(0.0d);
-        //this.hopperController.set(0.0d);
-        this.stopHopper();
-        this.stopConveyor();
+
+
+    @Override
+    public void unload() {
+        this.state = FeederState.UNLOADING;
     }
 
-    public void reverseFeeder() {
-        this.beltController.set(0.6d);
-        this.hopperController.set(-0.8d);
-        // TODO see above
+    void enable(FeederComponent component, FeederDirection direction) throws RuntimeException{
+        switch(component) {
+            case HOPPER:
+                if(direction == FeederDirection.FORWARDS) {
+                    this.conveyorSubsystem.feed();
+                } else {
+                    this.conveyorSubsystem.unload();
+                }
+            case CONVEYOR:
+                if(direction == FeederDirection.FORWARDS){
+                    this.conveyorSubsystem.feed();
+                } else {
+                    this.conveyorSubsystem.unload();
+                }
+                break;
+            default:
+                throw new RuntimeException("Specified component does not exist!");
```

Comment:
Consider adding the name of the unknown component into the exception message as well, ie `"Specified component \"" + component.toString() + "\" does not exist!"`, so that if this ever does occur, it's even easier to debug the problem.